src/mono/wasm/host/BrowserHost.cs

```diff
@@ -88,29 +89,104 @@ private async Task RunAsync(ILoggerFactory loggerFactory, CancellationToken toke
+        (ServerURLs serverURLs, IWebHost host) = await WebServer.StartAsync(webServerOptions, _logger, token);
         return (serverURLs, host);
     }
 
+    private static WebServerOptions CreateWebServerOptions(string[] urls, string appPath, Func<WebSocket, Task>? onConsoleConnected) => new
+    (
+        OnConsoleConnected: onConsoleConnected,
+        ContentRootPath: Path.GetFullPath(appPath),
+        WebServerUseCors: true,
+        WebServerUseCrossOriginPolicy: true,
+        Urls: urls
+    );
+
+    private static bool TryCreateDevServerOptions(BrowserArguments args, string[] urls, Func<WebSocket, Task>? onConsoleConnected, [NotNullWhen(true)] out DevServerOptions? devServerOptions)
+    {
+        const string staticWebAssetsV1Extension = ".StaticWebAssets.xml";
+        const string staticWebAssetsV2Extension = ".staticwebassets.runtime.json";
+
+        devServerOptions = null;
+
+        string targetDirectory = GetTargetDirectory(args.CommonConfig.RuntimeConfigPath);
+        if (args.CommonConfig.HostProperties.MainAssembly != null)
+        {
+            // If we have main assembly name, try to find static web assets manifest by precise name.
+
+            var mainAssemblyPath = Path.Combine(targetDirectory, args.CommonConfig.HostProperties.MainAssembly);
+            var staticWebAssetsPath = Path.ChangeExtension(mainAssemblyPath, staticWebAssetsV2Extension);
+            if (File.Exists(staticWebAssetsPath))
+            {
+                devServerOptions = CreateDevServerOptions(urls, staticWebAssetsPath, onConsoleConnected);
+            }
+            else
+            {
+                staticWebAssetsPath = Path.ChangeExtension(mainAssemblyPath, staticWebAssetsV1Extension);
+                if (File.Exists(staticWebAssetsPath))
+                    devServerOptions = CreateDevServerOptions(urls, staticWebAssetsPath, onConsoleConnected);
+            }
+        }
+        else
+        {
+            // If we don't have main assembly name, try to find static web assets manifest by search in the directory.
+
+            var staticWebAssetsPath = FindFirstFileWithExtension(targetDirectory, staticWebAssetsV2Extension)
+                ?? FindFirstFileWithExtension(targetDirectory, staticWebAssetsV1Extension);
+
+            if (staticWebAssetsPath != null)
+                devServerOptions = CreateDevServerOptions(urls, staticWebAssetsPath, onConsoleConnected);
+        }
+
+        return devServerOptions != null;
+    }
+
+    private static DevServerOptions CreateDevServerOptions(string[] urls, string staticWebAssetsPath, Func<WebSocket, Task>? onConsoleConnected) => new
+    (
+        OnConsoleConnected: onConsoleConnected,
+        StaticWebAssetsPath: staticWebAssetsPath,
+        WebServerUseCors: true,
+        WebServerUseCrossOriginPolicy: true,
+        Urls: urls
+    );
+
+    private static string GetTargetDirectory(string? runtimeConfigPath)
```

Comment:
nit: better as a local function in `TryCreateDevServerOptions`, since `target directory`.